core/src/main/java/org/infinispan/notifications/Listener.java

```diff
@@ -9,87 +9,169 @@
- * org.infinispan.notifications.cachelistener.annotation.CacheEntryPassivated}</td> <td valign="top">{@link
- * org.infinispan.notifications.cachelistener.event.CacheEntryPassivatedEvent}</td> <td valign="top">One or more cache entries were
- * passivated</td> </tr> <tr> <td valign="top">{@link org.infinispan.notifications.cachemanagerlistener.annotation.ViewChanged}</td>
- * <td valign="top">{@link org.infinispan.notifications.cachemanagerlistener.event.ViewChangedEvent}</td> <td
- * valign="top">A view change event was detected</td> </tr> <tr> <td valign="top">{@link
- * org.infinispan.notifications.cachelistener.annotation.TransactionRegistered}</td> <td valign="top">{@link
- * org.infinispan.notifications.cachelistener.event.TransactionRegisteredEvent}</td> <td valign="top">The cache has started
- * to participate in a transaction</td> </tr> <tr> <td valign="top">{@link org.infinispan.notifications.cachelistener.annotation.TransactionCompleted}</td>
- * <td valign="top">{@link org.infinispan.notifications.cachelistener.event.TransactionCompletedEvent}</td> <td
- * valign="top">The cache has completed its participation in a transaction</td> </tr> <tr> <td valign="top">{@link
- * org.infinispan.notifications.cachelistener.annotation.CacheEntryInvalidated}</td> <td valign="top">{@link
- * org.infinispan.notifications.cachelistener.event.CacheEntryInvalidatedEvent}</td> <td valign="top">A cache entry was
- * invalidated by a remote cache.  Only if cache mode is INVALIDATION_SYNC or INVALIDATION_ASYNC.</td> </tr>
+ * this.
+ * <h4>Listener Modes</h4>
+ * A listener can be configured to run in two different modes: sync or async.
+ * <p>The first, non-blocking, is a mode where the listener is notified in the invoking thread. Operations in this mode
+ * should be used when either the listener operation is expected to complete extremely fast or when the operation can be
+ * performed in a non-blocking by optionally returning a CompletionStage given back to the notification system to delay
+ * the operation until the stage is complete. This mode is the default mode, overrided by the {@link Listener#sync()}
+ * property. A method is non blocking if it declares that it returns a {@link java.util.concurrent.CompletionStage} or
+ * one of its subtypes. Note that the stage can return a value, but it will not be utilized. The user <b>must</b> be very
+ * careful that no blocking or long running operation is done while in a sync listener as it can cause possible thread
+ * starvation, depending on the calling thread. It is entirely acceptable to use your own thread pool to execute a
+ * listener operation and return a {@link java.util.concurrent.CompletionStage} signifying when it is complete.
+ * <p>The second, async, is pretty much identical to sync except that the original operation can continue and complete
+ * while the listener is notified in a different thread. Listeners that throw exceptions are always logged and are not
```

Comment:
I don't think we are ready to define what happens with sync yet. The behavior is not consistent at all.

Reply:
https://issues.jboss.org/browse/ISPN-3643

Reply:
It's weird that it's not documented in the Listener interface, but our tests definitely assume that a listener can veto a change by throwing an exception (e.g. `ListenerExceptionTest`). The problematic part is that they also assume an exception from a post listener to be propagated back to the caller, which means a tx may apply only a part of its changes.

Partial application of changes is also possible in non-tx or tx 1PC caches if the listener is not installed on all the nodes, or if the listener behaviour depends on more than just the event itself. Because of this I would always recommend wrapping the cache for validation instead of listeners, and I would be more than ok with removing the veto option from listeners /cc @tristantarrant @karesti